### PR TITLE
Porting DotCover module

### PR DESCRIPTION
**Pull request by cthangn**
@matthid Any suggestion with those `internal buildXXXXArgs` from other modules which are used in DotCover?

### CODE REVIEW

**Comment by matthid** [comment]
> internal buildXXXXArgs from other modules which are used in DotCover?

Not really, but somehow it looks wrong ;)

Would it be enough to add another public API to those modules with the following signature:

```fsharp
let buildArgs (setDotCoverParams: DotCoverParams -> DotCoverParams) : string =
   // ...
```

?

**Comment by cthangn** [comment]
It is not sufficient, because those `buildArgs` are dependent on those modules parameters not on `DotCoverParams`. It also doesn't feel right to have `DotCover` as dependency in those modules. I'd rather make those public than this. :) I may miss something though.

**Comment by vbfox** [comment]
I think that having `buildArgs` methods in each module with their current signature is the best that can be done.

Maybe with `/// [omit]` because it's not something that users need to find (Except if they wrap another coverage tool).

**Comment by matthid** [comment]
Ah yes, sorry this is actually what I meant... Basically we can make these things public as you suggest in the PR but we should make this a "feature" and unify the signatures (just like we did for `<Module>.run` we now have `<Module>.buildArgs`).

> Maybe with /// [omit] because it's not something that users need to find

I'm not 100% sure about this because sometimes it can be used for unblocking people, the disadvantage is that there is less need to contribute back ;)

File: src/app/Fake.DotNet.Testing.DotCover/DotCover.fs [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+open Fake.DotNet.Testing.XUnit2
+
+type DotCoverReportType = 
+  | Html = 0
+  | Json = 1
+  | Xml = 2
+  | NDependXml = 3
+
+/// The dotCover parameter type for running coverage
+[<CLIMutable>]
```
Comment (matthid) [inline src/app/Fake.DotNet.Testing.DotCover/DotCover.fs]: Can we remove `CLIMutable`?

File: src/app/Fake.DotNet.Testing.DotCover/DotCover.fs [hunk 2]
```diff
@@ -0,0 +1,334 @@
...
+      TargetExecutable = ""
+      TargetArguments = ""
+      TargetWorkingDir = ""
+      Filters = ""
+      AttributeFilters = ""
+      Output = "dotCoverSnapshot.dcvr"
+      CustomParameters = "" 
+      ErrorLevel = ErrorLevel.Error} 
+
+[<CLIMutable>]
```
Comment (matthid) [inline src/app/Fake.DotNet.Testing.DotCover/DotCover.fs]: here as well

File: src/app/Fake.DotNet.Testing.DotCover/DotCover.fs [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+
+let internal DotCoverMergeDefaults =
+     { ToolPath = Fake.IO.Globbing.Tools.findToolInSubPath "dotCover.exe" ((Path.GetFullPath ".") @@ "tools" @@ "DotCover")
+       WorkingDir = ""
+       Source = []
+       Output = "dotCoverSnapshot.dcvr"
+       TempDir = ""
+       CustomParameters = "" }
+
+[<CLIMutable>]
```
Comment (matthid) [inline src/app/Fake.DotNet.Testing.DotCover/DotCover.fs]: and one more

File: src/app/Fake.DotNet.Testing.DotCover/DotCover.fs [hunk 4]
```diff
@@ -0,0 +1,334 @@
...
+        failwithf "Error running %s with exitcode %d" toolPath result
+    else 
+        Trace.trace (sprintf "DotCover exited successfully")
+
+/// Runs the dotCover "cover" command, using a target executable (such as NUnit or MSpec) and generates a snapshot file.
+///
+/// ## Parameters
+///
+///  - `setParams` - Function used to overwrite the dotCover default parameters.
+let DotCover (setParams: DotCoverParams -> DotCoverParams) =
```
Comment (matthid) [inline src/app/Fake.DotNet.Testing.DotCover/DotCover.fs]: Can we rename this to `run` as other modules?

File: src/app/Fake.DotNet.Testing.DotCover/DotCover.fs [hunk 5]
```diff
@@ -0,0 +1,334 @@
...
+///
+///  - `setParams` - Function used to overwrite the dotCover report default parameters.
+///
+/// ## Sample
+///
+///     DotCoverReport (fun p -> { p with 
+///                         Source = artifactsDir @@ "dotCoverSnapshot.dcvr"
+///                         Output = artifactsDir @@ "dotCoverReport.xml"
+///                         ReportType = DotCoverReportType.Xml })
+let DotCoverReport (setParams: DotCoverReportParams -> DotCoverReportParams) =
```
Comment (matthid) [inline src/app/Fake.DotNet.Testing.DotCover/DotCover.fs]: `DotCover.report`?

File: src/app/Fake.DotNet.Testing.DotCover/DotCover.fs [hunk 6]
```diff
@@ -0,0 +1,334 @@
...
+///
+/// ## Sample
+///
+///     !! (buildDir @@ buildMode @@ "/*.Unit.Tests.dll") 
+///         |> DotCoverNUnit 
+///             (fun dotCoverOptions -> { dotCoverOptions with 
+///                     Output = artifactsDir @@ "NUnitDotCoverSnapshot.dcvr" }) 
+///             (fun nUnitOptions -> { nUnitOptions with
+///                     DisableShadowCopy = true })
+let DotCoverNUnit (setDotCoverParams: DotCoverParams -> DotCoverParams) (setNUnitParams: NUnit.Common.NUnitParams -> NUnit.Common.NUnitParams) (assemblies: string seq) =
```
Comment (matthid) [inline src/app/Fake.DotNet.Testing.DotCover/DotCover.fs]: `DotCover.runNunit`?

**Comment by matthid** [comment]
Sorry for the late review

**Pull request merged**